apollo-router/src/protocols/multipart.rs

```diff
@@ -0,0 +1,208 @@
+            return Poll::Ready(None);
+        }
+        match self.stream.as_mut().poll_next(cx) {
+            Poll::Ready(message) => match message {
+                Some(None) => {
+                    // It's the ticker for heartbeat for subscription
+                    let buf = if self.is_first_chunk {
+                        self.is_first_chunk = false;
+                        Bytes::from_static(
+                            &b"\r\n--graphql\r\ncontent-type: application/json\r\n\r\n{}\r\n--graphql\r\n"[..]
+                        )
+                    } else {
+                        Bytes::from_static(
+                            &b"content-type: application/json\r\n\r\n{}\r\n--graphql\r\n"[..],
+                        )
+                    };
+
+                    Poll::Ready(Some(Ok(buf)))
+                }
+                Some(Some(mut response)) => {
+                    let mut buf = if self.is_first_chunk {
+                        self.is_first_chunk = false;
+                        Vec::from(&b"\r\n--graphql\r\ncontent-type: application/json\r\n\r\n"[..])
+                    } else {
+                        Vec::from(&b"content-type: application/json\r\n\r\n"[..])
+                    };
+                    let is_still_open =
+                        response.has_next.unwrap_or(false) || response.subscribed.unwrap_or(false);
+                    match self.mode {
+                        ProtocolMode::Subscription => {
+                            let resp = SubscriptionPayload {
+                                errors: if is_still_open {
+                                    Vec::new()
+                                } else {
+                                    response.errors.drain(..).collect()
+                                },
+                                payload: match response.data {
+                                    None | Some(Value::Null) => None,
+                                    _ => response.into(),
+                                },
+                            };
+
+                            serde_json::to_writer(&mut buf, &resp)?;
+                        }
+                        ProtocolMode::Defer => {
+                            serde_json::to_writer(&mut buf, &response)?;
+                        }
+                    }
+
+                    if is_still_open {
+                        buf.extend_from_slice(b"\r\n--graphql\r\n");
+                    } else {
+                        self.is_terminated = true;
+                        buf.extend_from_slice(b"\r\n--graphql--\r\n");
+                    }
+
+                    Poll::Ready(Some(Ok(buf.into())))
+                }
+                None => Poll::Ready(None),
+            },
+            Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use futures::stream;
+    use serde_json_bytes::ByteString;
+
+    use super::*;
+
+    // TODO add test with empty stream
```

Comment:
Do we have a ticket for follow-up here?

Reply:
https://github.com/apollographql/router/issues/3293